test/test_testing.py

```diff
@@ -7329,3 +7329,81 @@ def test_scenario_available():
     ctx = ops.testing.Context(ops.CharmBase, meta={'name': 'foo'})
     state = ctx.run(ctx.on.start(), ops.testing.State())
     assert isinstance(state.unit_status, ops.testing.UnknownStatus)
+
+
+@pytest.mark.parametrize('test_context', ['init', 'event'])
+@pytest.mark.parametrize(
+    'is_leader', [pytest.param(True, id='leader'), pytest.param(False, id='minion')]
+)
+def test_relation_validates_access(
+    request: pytest.FixtureRequest, is_leader: bool, test_context: str
+):
+    """Test that relation databag read/write access in __init__ is the same as in observers."""
+    APP_NAME = 'charm'
+    REL_NAME = 'relation'
+    ACTION_NAME = 'action'
+    KEY = 'key'
+    LOCAL_VAL = 'local value'
+    REMOTE_VAL = 'remote value'
+
+    class Charm(ops.CharmBase):
+        def __init__(self, framework: ops.Framework):
+            super().__init__(framework)
+            framework.observe(self.on[ACTION_NAME].action, self._on_action)
+            self.validated = 0
+            self.test_validation('init')
+
+        def _on_action(self, action: ops.ActionEvent):
+            self.test_validation('event')
+
+        def test_validation(self, context: str):
+            if context != test_context:
+                return
+            self.validated += 1
+            rel = self.model.get_relation(REL_NAME)
+            assert rel is not None
+
+            # remote application databag
+            # any unit can read the remote application databag
+            remote_app_data = rel.data[rel.app]
+            assert remote_app_data[KEY] == REMOTE_VAL
+            assert len(remote_app_data.items()) == 1
+            # no unit can write to the remote application databag
+            with pytest.raises(ops.RelationDataAccessError):
+                remote_app_data[KEY] = 'something'
+
+            # local application databag
+            local_app_data = rel.data[self.app]
+            # only the leader can read or write the local application databag
+            if self.unit.is_leader():
+                assert local_app_data[KEY] == LOCAL_VAL  # test read
+                local_app_data[KEY] = 'new value'  # test write
+            else:
+                with pytest.raises(ops.RelationDataAccessError):
+                    local_app_data[KEY]
+            # these probably fail at real runtime with a ModelError
+            # but pass here because the validation methods are only hooked up to get/set
+            assert len(local_app_data.items()) == 1
+            assert KEY in local_app_data
+
+    harness = ops.testing.Harness(
+        Charm,
+        meta=f"""
+name: {APP_NAME}
+requires:
+  {REL_NAME}:
+    interface: interface-name
+""",
+        actions=f"""
+{ACTION_NAME}:
```

Comment:
I find that such constants in tests usually obscure the meaning, for example here, where just putting the constants inline would be clearer. The fact that they're repeated a couple of times isn't a big deal -- the scope is small, and if you get one side wrong the test will fail.